plasma_framework/contracts/src/exits/payment/controllers/PaymentChallengeIFEInputSpent.sol

```diff
@@ -0,0 +1,129 @@
+        PaymentInFlightExitRouterArgs.ChallengeInputSpentArgs args;
+        PaymentExitDataModel.InFlightExit ife;
+    }
+
+    function buildController(
+        PlasmaFramework framework,
+        SpendingConditionRegistry spendingConditionRegistry,
+        uint256 supportedTxType
+    )
+        public
+        view
+        returns (Controller memory)
+    {
+        return Controller({
+            framework: framework,
+            isDeposit: IsDeposit.Predicate(framework.CHILD_BLOCK_INTERVAL()),
+            spendingConditionRegistry: spendingConditionRegistry,
+            supportedTxType: supportedTxType
+        });
+    }
+
+    function run(
+        Controller memory self,
+        PaymentExitDataModel.InFlightExitMap storage inFlightExitMap,
+        PaymentInFlightExitRouterArgs.ChallengeInputSpentArgs memory args
+    )
+        public
+    {
+        uint192 exitId = ExitId.getInFlightExitId(args.inFlightTx);
+        PaymentExitDataModel.InFlightExit storage ife = inFlightExitMap.exits[exitId];
+
+        require(ife.exitStartTimestamp != 0, "In-flight exit doesn't exist");
+        require(ife.isInputPiggybacked(args.inFlightTxInputIndex), "The indexed input has not been piggybacked");
+
+        require(
+            keccak256(args.inFlightTx) != keccak256(args.spendingTx),
+            "The spending transaction is the same as the in-flight transaction"
+        );
+
+        ChallengeIFEData memory data = ChallengeIFEData({
+            controller: self,
+            args: args,
+            ife: inFlightExitMap.exits[exitId]
+        });
+
+        verifySpentInputEqualsIFEInput(data);
+
+        verifySpendingCondition(data);
+
+        // Remove the input from the piggyback map
+        ife.clearInputPiggybacked(args.inFlightTxInputIndex);
+
+        // Pay out the bond.
+        msg.sender.transfer(PIGGYBACK_BOND);
+
+        emit InFlightExitInputBlocked(msg.sender, keccak256(args.inFlightTx), args.inFlightTxInputIndex);
+    }
+
+    function verifySpentInputEqualsIFEInput(ChallengeIFEData memory data) private pure {
+        bytes32 ifeInputOutputId = data.ife.inputs[data.args.inFlightTxInputIndex].outputId;
+
+        UtxoPosLib.UtxoPos memory utxoPos = UtxoPosLib.build(TxPosLib.TxPos(data.args.inputTxPos), data.args.inputTxOutputIndex);
+        bytes32 spendingTxInputOutputId = data.controller.isDeposit.test(utxoPos.blockNum())
+                ? OutputId.computeDepositOutputId(data.args.inputTx, utxoPos.outputIndex(), utxoPos.value)
+                : OutputId.computeNormalOutputId(data.args.inputTx, utxoPos.outputIndex());
+
+        require(ifeInputOutputId == spendingTxInputOutputId, "Spent input is not the same as piggybacked input");
+    }
+
+    function verifySpendingCondition(ChallengeIFEData memory data) private view {
+        ISpendingCondition condition = data.controller.spendingConditionRegistry.spendingConditions(
+            data.args.spendingTxInputOutputType, data.args.spendingTxType
```

Comment:
We need to be careful here. I think we should verify that output type and tx type from args match spending tx type and output type of challenged input.

Reply:
Ok. To verify the output type, we need to pass in the outputGuardPreimage and check OutputGuardHandler.isValid(), right?